subt/octomap.py

```diff
@@ -154,10 +154,27 @@ def frontiers(img, start, draw=False):
         plt.axes().set_aspect('equal', 'datalim')
         plt.show()
 
-    driveable = white[:1023, :1023] | white[1:, :1023] | white[1:, 1:] | white[:1023, 1:]
+    driveable = white[:, :]
+
+    # use 1 pixel surounding
+    d = driveable[:, :]
+    d2 = d[2:, :] & d[1:-1, :] & d[:-2, :]
+    d3 = d2[:, 2:] & d2[:, 1:-1] & d2[:, :-2]
+    z = np.zeros((1022, 1), dtype=np.bool)
+    d = np.hstack([z, d3, z])
+    z = np.zeros((1, 1024), dtype=np.bool)
+    driveable = np.vstack([z, d, z])
+
+    img[driveable, 0] = 128  # gray
+    img[driveable, 1] = 128
+    img[driveable, 2] = 128
+
     i = np.argmax(score)
     limit_score = 3*max(score)/4
-    goals = [(xy[1][i], xy[0][i]) for i in range(len(xy[0])) if score[i] > limit_score]
+    goals = []
+    for dx in range(-1, 2):
+        for dy in range(-1, 2):
+            goals.extend([(xy[1][i] + dx, xy[0][i] + dy) for i in range(len(xy[0])) if score[i] > limit_score])
```

Comment:
For loop in a For loop in a For loop ... in Python ... Let's find a better way.

(this is a change I would like to see)
First, is seems that at least after plotting, you do not care about `xy` columns with low score. So let's drop them beforehand with numpy:
`xy = xy[:, score > limit_score]`
That should let you:
a) `goals.append(xy + np.repeat(np.asarray([[dy], [dx]]), xy.shape[1], axis=1))`
b) after the for loops, `goals = np.hstack(goals).T[:,::-1]`
You will end up with NumGoals x 2 sized array, which imo `find_path` should be able to accept instead of the current list of two-element lists.

(this is for entertainment, not a desired change)
I am also convinced that by turning `xy` into a three-dimensional 9x copy of itself and preparing the (dy, dx) combinations into a nice 2x1x9 tensor, you could do even the outer two loops inside numpy and do a simple `.reshape((2, -1))` on the result. But that probably goes too far away from readable code, so unless we hit a performance problem somewhere around here, let's not do that.

Reply:
Uhh, I should use this as hint for the "numpy version" - yeah, I am aware that Python for-loops are too much even in this experiment. I will check this when I am fresher, thanks a lot!

Reply:
The numpy version would not be easy for me even when I am fresh. Even when writing the comment it took me time to convince myself that I got the correct dimensions that make sense. And maybe I didn't. I would go with the simpler version first and maybe the other one in a standalone follow-up.

Reply:
It is still strange for me ... one fix was easy:
```
    xy = xy[:, score > limit_score]
TypeError: tuple indices must be integers or slices, not tuple
```
fixed by `xy = np.array(xy)[:, score > limit_score]`. But the second part asserts (some older run, where I am trying to get 1:1) and it works only if I use `goals = [tuple(x) for x in goals.tolist()]` as extra "repacking". I hope to sort this out soon ...
p.s. thanks for mail reminder

Reply:
I am not sure if fix via `goals = list(map(tuple, goals))` would be nicer. The reason is that the planner expects tuple pairs `(x, y)` ... but I do not tweak path planner in this PR. The goal can be as well marked in the image.

Reply:
p.s. changed to `set()` and hopefully ready for merge